src/inspector_profiler.cc

```diff
@@ -503,6 +507,217 @@ static void StopCoverage(const FunctionCallbackInfo<Value>& args) {
+    CHECK_EQ(0, rc);
+  }
+}
+
+void ProfileWatchdog::Stop() {
+  if (state() != ProfileWatchdogState::kClosing &&
+      state() != ProfileWatchdogState::kClosed) {
+    set_state(ProfileWatchdogState::kClosing);
+  }
+}
+
+void ProfileWatchdog::Start(const FunctionCallbackInfo<Value>& args) {
+  ProfileWatchdog* watchdog = Unwrap<ProfileWatchdog>(args.Holder());
+  if (watchdog) {
+    watchdog->Start();
+  }
+}
+
+void ProfileWatchdog::Stop(const FunctionCallbackInfo<Value>& args) {
+  ProfileWatchdog* watchdog = Unwrap<ProfileWatchdog>(args.Holder());
+  if (watchdog) {
+    watchdog->Stop();
+  }
+}
+
+void ProfileWatchdog::Run(void* arg) {
+  ProfileWatchdog* wd = static_cast<ProfileWatchdog*>(arg);
+  uv_run(&wd->loop_, UV_RUN_DEFAULT);
+}
+
+void ProfileWatchdog::Timer(uv_timer_t* timer) {
+  ProfileWatchdog* w = ContainerOf(&ProfileWatchdog::timer_, timer);
+  Environment* env = w->env();
+  if (!env || env->is_stopping()) {
+    return;
+  }
+  if (w->state() == ProfileWatchdogState::kClosed) {
+    return;
+  }
+  env->RequestInterrupt([watchdog = std::move(w)](Environment* env) {
+    if (env->is_stopping() ||
+        watchdog->state() == ProfileWatchdogState::kClosed) {
+      return;
+    }
+    if (watchdog->state() == ProfileWatchdogState::kClosing) {
+      delete watchdog;
+      return;
+    }
+    if (watchdog->TimeoutHandler()) {
+      watchdog->AddTask(
+          [watchdog = std::move(watchdog)]() { watchdog->SetTimeout(); });
+    }
+  });
+}
+
+template <typename Fn>
+void ProfileWatchdog::AddTask(Fn&& cb, CallbackFlags::Flags flags) {
+  auto callback = tasks_.CreateCallback(std::move(cb), flags);
+  {
+    Mutex::ScopedLock lock(task_mutex_);
+    tasks_.Push(std::move(callback));
+  }
+  uv_async_send(&async_);
+}
+
+void ProfileWatchdog::HandleTasks() {
+  while (tasks_.size() > 0) {
+    CallbackQueue<void> queue;
+    {
+      Mutex::ScopedLock lock(task_mutex_);
+      queue.ConcatMove(std::move(tasks_));
+    }
+    while (auto head = queue.Shift()) head->Call();
+  }
+}
+
+void ProfileWatchdog::SetTimeout() {
+  int rc = uv_timer_start(&timer_, &ProfileWatchdog::Timer, interval_, 0);
+  CHECK_EQ(0, rc);
+}
+
+Local<FunctionTemplate> ProfileWatchdog::GetConstructorTemplate(
+    Environment* env) {
+  Local<FunctionTemplate> tmpl = env->profile_watchdog_ctor_template();
+  if (tmpl.IsEmpty()) {
+    Isolate* isolate = env->isolate();
+    tmpl = NewFunctionTemplate(isolate, nullptr);
+    tmpl->SetClassName(
+        FIXED_ONE_BYTE_STRING(env->isolate(), "ProfileWatchdog"));
+    SetProtoMethod(isolate, tmpl, "start", ProfileWatchdog::Start);
+    SetProtoMethod(isolate, tmpl, "stop", ProfileWatchdog::Stop);
+    env->set_profile_watchdog_ctor_template(tmpl);
+  }
+  return tmpl;
+}
+
+ProfileWatchdog::ProfileWatchdog(Environment* env, Local<Object> object)
+    : BaseObject(env, object) {
+  int rc;
+
+  rc = uv_loop_init(&loop_);
+  CHECK_EQ(0, rc);
```

Comment:
uv_loop_init() can (and sometimes will) fail. It's not nice to crash the process when that happens.

Reply:
I have no idea about this. I find the `node_watchdog.cc` will throw a fatal error when `uv_loop_init` fails.
```
rc = uv_loop_init(&loop_);
if (rc != 0) {
  FatalError("node::Watchdog::Watchdog()",
             "Failed to initialize uv loop.");
}
```
or `TCPWrap::TCPWrap` in `tcp_wrap.cc` will crash the process when `uv_tcp_init` fails.

Reply:
Maybe we can call libuv function in `ProfileWatchdog::Start` function and throw a error to JS if libuv function fails.

Reply:
That would be better. Even FatalError() is already a step up but something the user can handle is better still.